src/core/namespace.c

```diff
@@ -1911,43 +1982,42 @@ int open_netns_path(const int netns_storage_socket[static 2], const char *path)
          * it. This is supposed to be called ahead of time, i.e. before setup_netns() which will allocate a
          * new anonymous netns if needed. */
 
-        if (lockf(netns_storage_socket[0], F_LOCK, 0) < 0)
-                return -errno;
+        r = lockfp(netns_storage_socket[0], &storage_socket0_lock);
+        if (r < 0)
+                return r;
 
         netns = receive_one_fd(netns_storage_socket[0], MSG_DONTWAIT);
         if (netns == -EAGAIN) {
                 /* Nothing stored yet. Open the file from the file system. */
-
                 netns = open(path, O_RDONLY|O_NOCTTY|O_CLOEXEC);
                 if (netns < 0) {
-                        r = -errno;
-                        goto fail;
-                }
-
-                r = fd_is_network_ns(netns);
-                if (r == 0) { /* Not a netns? Refuse early. */
-                        r = -EINVAL;
-                        goto fail;
+                        /* Not exists yet. Let's create a new namespace. */
+                        netns = netns_create(path, true);
+                        if (netns < 0)
+                                return netns;
+                } else {
+                        r = fd_is_network_ns(netns);
+                        if (r < 0 && r != -EUCLEAN) /* EUCLEAN: we don't know */
+                                return r;
+                        if (r == 0) {
+                                /* File exists, but netns is not mounted on the path. */
+                                netns = netns_create(path, false);
```

Comment:
Are we sure we want to have this behaviour? Maybe its better to just hard-fail when the file pointed at by `NetworkNamespacePath` is not a network namespace. This approach feels like it could hide subtle errors.

Reply:
OK. Dropped.